atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/CqlClientImpl.java

```diff
@@ -124,7 +125,10 @@ public void close() {
         Statement executableStatement = querySpec.makeExecutableStatement(statement)
                 .setConsistencyLevel(querySpec.queryConsistency());
 
-        return execute(executableStatement, querySpec.executor(), querySpec.rowStreamAccumulatorFactory().get());
+        return execute(
+                executableStatement,
+                querySpec.cqlQueryContext().executor(),
+                querySpec.rowStreamAccumulatorFactory().get());
     }
 
     private <V> ListenableFuture<V> execute(
```

Comment:
So how about this for a threading model: 

1. We do the iteration and accumulation on the Cassandra threads - this code should be super simple and super fast, expecially if we specialize to LIMIT 1 queries for now.
2. (maybe, but might not be worth it to start with) Remove all the synchronization and threadsafety stuff from RowStreamAccumulator implementations: for this to work we simply need to double check that the driver only ever fetches pages serially and only ever returns control serially. Then just make sure that memory visibility is kept.
3. Then the only thing that happens on the special threadpool is AsyncCellLoader's Futures.whenAllSucceed: and memory visibility should be there nicely, because all the futures implementations we use are threadsafe so everything should be propagated correctly.

Reply:
1. Iteration and accumulation have been resolved by a commit from @jkozlowski 

2. Removed synchronization and thread safety. Each page of a query is submitted for processing only after the current page is processed and since an accumulator is on a per query basis there is not possibility for concurrent access of two threads to the same accumulator. 

3. However we have no guarantees about the visibility of the results in the `AsyncCellLoder` collection phase so internal fields of `GetQueryAccumulator` are `volatile`.

Reply:
3. There should be guarantee, because futures are synchronized and thread-safe by default. So there should be a happens-before for any stuff that happens.